master/buildbot/db/migrate/versions/056_add_test_result_tables.py

```diff
@@ -0,0 +1,163 @@
+# This file is part of Buildbot. Buildbot is free software: you can
+# redistribute it and/or modify it under the terms of the GNU General Public
+# License as published by the Free Software Foundation, version 2.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+
+
+import sqlalchemy as sa
+
+from buildbot.util import sautils
+
+
+def upgrade(migrate_engine):
+    metadata = sa.MetaData()
+    metadata.bind = migrate_engine
+
+    sautils.Table(
+        'builds', metadata,
+        sa.Column('id', sa.Integer, primary_key=True),
+        # ...
+    )
+
+    sautils.Table(
+        'builders', metadata,
+        sa.Column('id', sa.Integer, primary_key=True),
+        # ...
+    )
+
+    steps = sautils.Table(
+        'steps', metadata,
+        sa.Column('id', sa.Integer, primary_key=True),
+        # ...
+    )
+
+    test_result_sets = sautils.Table(
+        'test_result_sets', metadata,
+
+        sa.Column('id', sa.Integer, primary_key=True),
+
+        sa.Column('builderid', sa.Integer,
+                  sa.ForeignKey('builders.id', ondelete='CASCADE'),
+                  nullable=False),
+
+        sa.Column('buildid', sa.Integer,
```

Comment:
lets document here the reason why we want to denormalize the data model.

Reply:
It's documented, see https://github.com/buildbot/buildbot/pull/5183/files#diff-58b3908894b0546af761475a0243ffd1R622